### PR TITLE
JENKINS-28600: Support Dockerfiles with non-standard names.

### PR DESCRIPTION
**Pull request by johndistasio**
Hello,

This pull request addresses JENKINS-28600 by changing the build process to work a little more like it does with the Docker client.

### CODE REVIEW

File: src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java [hunk 1]
```diff
@@ -295,17 +306,17 @@ private boolean maybeTagOnly() throws MacroEvaluationException, IOException, Int
             }
             return executeCmd(result);
         }
-        
+
 		private boolean buildAndTag() throws MacroEvaluationException, IOException, InterruptedException {
-			String context = defined(getDockerfilePath()) ?
-					getDockerfilePath() : ".";
+			File context = new File(defined(getBuildContext()) ? getBuildContext() : ".");
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java]: Should use <code>FilePath</code> and default to the workspace IMO

File: src/main/resources/com/cloudbees/dockerpublish/DockerBuilder/config.jelly [hunk 2]
```diff
@@ -51,8 +51,13 @@
...
 
-    <f:entry title="Directory Dockerfile is in" field="dockerfilePath"
-      description="The project root is where the Dockerfile is looked for by default - if you need another path, enter it here">
+    <f:entry title="Build Context" field="buildContext"
+      description="The project root path for the build. Defaults to the workspace root if not specified.">
+      <f:textbox />
+    </f:entry>
+
+    <f:entry title="Dockerfile Path" field="dockerfilePath"
+      description="Path to the Dockerfile for this build. Can be used to specify a Dockerfile with a non-standard filename. Uses Docker client default if not specified.">
```
Comment (oleg-nenashev) [inline src/main/resources/com/cloudbees/dockerpublish/DockerBuilder/config.jelly]: You should mention the path is relative to <code>buildContext</code>

File: src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java [hunk 3]
```diff
@@ -126,6 +128,15 @@ public void setForcePull(boolean forcePull) {
         this.forcePull = forcePull;
     }
 
+    public String getBuildContext() {
+        return buildContext;
+    }
+
+    @DataBoundSetter
+    public void setBuildContext(String buildContext) {
+        this.buildContext = buildContext;
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java]: Should be optional => <code>@ChechForNull</code>

**Comment by oleg-nenashev** [comment]
Seems this behavior will be a breaking change for builds using absolute paths. It needs to be updated

**Comment by zeeraw** [comment]
@johndistasio any ideas?

**Comment by johndistasio** [comment]
@zeeraw Yeah, I plan on addressing the concerns when I get more than a few minutes to work on this stuff.

**Comment by johndistasio** [comment]
@oleg-nenashev I made some changes to address the issues you pointed out, please take a look when you get a chance.

File: src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java]: Trimming would be also useful. See methods in hudson.Util

**Comment by oleg-nenashev** [comment]
@johndistasio 
Thanks for your efforts. I like the current implementation, and hopefully we will get it merged soon.

There's only one minor comment left (regarding the input data). It would be also useful to add the new field to roundtrip unit tests.

**Comment by johndistasio** [comment]
Added the requested input cleanup. I'll look into adding the round-trip testing.

File: src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java [hunk 4]
```diff
@@ -295,17 +309,18 @@ private boolean maybeTagOnly() throws MacroEvaluationException, IOException, Int
             }
             return executeCmd(result);
         }
-        
+
 		private boolean buildAndTag() throws MacroEvaluationException, IOException, InterruptedException {
-			String context = defined(getDockerfilePath()) ?
-					getDockerfilePath() : ".";
+			FilePath context = defined(getBuildContext()) ?
+                new FilePath(new File(getBuildContext())) : build.getWorkspace();
```
Comment (oleg-nenashev) [inline src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java]: minor: spaces instead of tabs

**Comment by mad0house** [comment]
Is this ever going to be merged in?

**Comment by oleg-nenashev** [comment]
:+1: from me, but I'm not the plugin owner
Cc @carlossg and @michaelneale

**Comment by michaelneale** [comment]
LGTM (Carlos knows code more recently)

**Comment by carlossg** [comment]
sorry for the delay

**Pull request merged**